### PR TITLE
Refactor FXIOS-7641 [v121] Refactor findAndHandle(route:)

### PR DESCRIPTION
**Pull request by mattreaganmozilla**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-7431)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/17043)

## :bulb: Description

Refactors `findAndHandle(route:)` to split the logic into separate steps:

1. Identify a valid `Coordinator` for a given `Route` 
2. Dismiss any children on target Coordinator (if needed)
3. Perform the actual `handle()` action

The important change is that the recursion of finding the coordinator is allowed to fully complete before we proceed to step 2. This provides a more robust solution for a few related issues ([FXIOS-7631](https://mozilla-hub.atlassian.net/browse/FXIOS-7631) & [FXIOS-7537](https://mozilla-hub.atlassian.net/browse/FXIOS-7537) etc).

This PR also removes the previous `newlyAdded` workaround for [FXIOS-7631](https://mozilla-hub.atlassian.net/browse/FXIOS-7631) since with these changes it is no longer needed.

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed I updated documentation / comments for complex code and public methods

### CODE REVIEW

**Comment by mattreaganmozilla** [comment]
@OrlaM @lmarceau Initial stab at these refactors. Am doing some regression testing of some of the related tickets, so far everything is working as expected but LMK if you have any concerns/thoughts.

File: Client/Coordinators/Browser/BrowserCoordinator.swift [hunk 1]
```diff
@@ -170,46 +170,51 @@ class BrowserCoordinator: BaseCoordinator,
             return false
         }
 
+        switch route {
+        case .searchQuery, .search, .searchURL, .glean, .homepanel, .action, .fxaSignIn, .defaultBrowser:
```
Comment (lmarceau) [inline Client/Coordinators/Browser/BrowserCoordinator.swift:174]: I think it misses the `.settings` route?

Comment (mattreaganmozilla) [reply Client/Coordinators/Browser/BrowserCoordinator.swift:174]: It's right below this, that one had some custom logic so it's the one route type which needs to check the associated value first (afaics).

Comment (lmarceau) [reply Client/Coordinators/Browser/BrowserCoordinator.swift:174]: Ah oups 🙈

**Review by OrlaM — APPROVED** [review]

**Pull request merged**